### PR TITLE
Substantially better error recovery for method args

### PR DESCRIPTION
**Pull request by jez**
### Motivation

This causes a huge source of pain when using Sorbet in large codebases, because
small edits frequently cause large swaths of a file to be thrown away.

### Test plan

See included automated tests.

### CODE REVIEW

File: parser/parser/cc/grammars/typedruby.ypp [hunk 1]
```diff
@@ -1512,6 +1512,15 @@ int maybe_inject_tNL_impl(parser::symbol_type &yyla, ruby_parser::typedruby27 &d
...
                     }
+                // There's an explanation of what's going on in this rule in the `assocs` rule.
+                | args tCOMMA fcall error assocs opt_block_arg
+                    {
+                      auto &args = $1;
+                      args->emplace_back(driver.build.call_method(self, nullptr, nullptr, $3, nullptr, nullptr, nullptr));
+                      args->emplace_back(driver.build.associate(self, nullptr, $5, nullptr));
+                      args->concat($6);
+                      $$ = args;
+                    }
```
Comment (elliottt) [inline parser/parser/cc/grammars/typedruby.ypp:1523]: Does this rule need to add an error?

File: parser/parser/cc/grammars/typedruby.ypp [hunk 2]
```diff
@@ -3676,12 +3695,62 @@ f_opt_paren_args: f_paren_args
...
                 | assocs tCOMMA assoc
                     {
                       auto &list = $1;
                       list->emplace_back($3);
                       $$ = list;
                     }
+                | assocs tCOMMA error
+                    {
+                      $$ = $1;
```
Comment (elliottt) [inline parser/parser/cc/grammars/typedruby.ypp:3715]: Should this rule add an error?

File: parser/parser/cc/grammars/typedruby.ypp [hunk 3]
```diff
@@ -3676,12 +3695,62 @@ f_opt_paren_args: f_paren_args
...
+                    }
+                | assocs tCOMMA tIDENTIFIER error
+                    {
+                      auto &list = $1;
+                      // TODO(jez) Return a special function call like `<kw__a` so that you can complete kwargs
+                      auto err = driver.build.call_method(self, nullptr, nullptr, $3, nullptr, nullptr, nullptr);
+                      list->emplace_back(driver.build.pair_keyword(self, $3, err));
+                      $$ = list;
+                    }
```
Comment (elliottt) [inline parser/parser/cc/grammars/typedruby.ypp:3724]: Should this rule add an error?

**Review by elliottt — CHANGES_REQUESTED** [review]
Looking forward to using this!

My one concern right now is that there are a fair number of productions that match `error` and don't explicitly add errors. Either we should add comments explaining why it's not necessary to do so in those cases, or explicitly add errors.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_KxxpXn6VKL0Did
→ https://go/builds/bui_Kxxpw5kIvgI7Ey

**Review by elliottt — APPROVED** [review]

**Comment by elliottt** [comment]
After discussion from a DM: I misunderstood how `error` works in bison, and assumed that it would eat the parse error like it does in [happy](https://www.haskell.org/happy/doc/html/sec-error.html). As errors will be raised in all these cases, all my previous comments don't apply 👍

**Pull request merged**